[trainer] deepspeed integration (#9211)
huggingface/transformers: docs/source/main_classes/trainer.rst

DeepSpeed
-----------------------------------------------------------------------------------------------------------------------

The other important third party component that this trainer supports is `DeepSpeed
<https://github.com/microsoft/DeepSpeed>`__.

It implements almost everything described in the `ZeRO paper <https://arxiv.org/abs/1910.02054>`__. As of this writing
it is still missing ZeRO's stage 3. "Parameter Partitioning (Pos+g+p )", but it fully supports:

1. Optimizer State Partitioning (stage 1)
2. Add Gradient Partitioning (stage 2)

To enable DeepSpeed you need to first install DeepSpeed following `the instructions
<https://github.com/microsoft/deepspeed#installation>`__.

And when the installation has been completed you need to adjust the command line arguments as following:

1. replace ``python -m torch.distributed.launch`` with ``deepspeed``
2. add a new argument ``--deepspeed ds_config.json`` where ``ds_config.json`` is the DeepSpeed configuration file as
   documented at https://www.deepspeed.ai/docs/config-json/

Therefore if your original program looked as following:

.. code-block:: bash

>     python -m torch.distributed.launch --nproc_per_node=2 your_program.py <your program\'s normal args>

Reviewer: I think you should discuss how to configure multi-gpu and multi-node runs with deepspeed. For example, what replaces --nproc_per_node when using deepspeed. A good pointer would be [this](https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node).

Author: That's super useful - thank you for the suggestion and the link to the resource, @tjruwase. 

I updated the doc.

I'm sure the doc will evolve as users will start asking questions.

p.s for some reason github doesn't ever show your username in the @ completion, even if I type half the first letters of it - very odd.

.. autoclass:: transformers.TFTrainingArguments
    :members:

Trainer Integrations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The Trainer has been extended to support libraries that may dramatically improve your training time and fit much bigger
models.

Currently it support third party solutions, `DeepSpeed <https://github.com/microsoft/DeepSpeed>`__ and `FairScale
<https://github.com/facebookresearch/fairscale/>`__, which implement parts of the paper `ZeRO: Memory Optimizations
Toward Training Trillion Parameter Models, by Samyam Rajbhandari, Jeff Rasley, Olatunji Ruwase, Yuxiong He
<https://arxiv.org/abs/1910.02054>`__.

This provided support is new and experimental as of this writing.

You will need at least 2 GPUs to benefit from these features.

FairScale
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> By integrating `FairScale <https://github.com/facebookresearch/fairscale/>`__ the Trainer provides support for the

Reviewer 2: I'm okay will all Trainer instances but since we're talking about the Trainer class, they should all be replaced to :class:`~transformers.Trainer`.

following features from `the ZeRO paper <https://arxiv.org/abs/1910.02054>`__:

1. Optimizer State Sharding
2. Gradient Sharding

To deploy this feature:

1. Install the library via pypi:

   .. code-block:: bash

       pip install fairscale

   or find more details on `the FairScale's github page
   <https://github.com/facebookresearch/fairscale/#installation>`__.

2. Add ``--sharded_ddp`` to the command line arguments, and make sure you have added the distributed launcher ``-m
   torch.distributed.launch --nproc_per_node=NUMBER_OF_GPUS_YOU_HAVE`` if you haven't been using it already.

For example here is how you could use it for ``finetune_trainer.py`` with 2 GPUs:

.. code-block:: bash

    cd examples/seq2seq
    python -m torch.distributed.launch --nproc_per_node=2 ./finetune_trainer.py \
    --model_name_or_path sshleifer/distill-mbart-en-ro-12-4 --data_dir wmt_en_ro \
    --output_dir output_dir --overwrite_output_dir \
    --do_train --n_train 500 --num_train_epochs 1 \
    --per_device_train_batch_size 1  --freeze_embeds \
    --src_lang en_XX --tgt_lang ro_RO --task translation \
    --fp16 --sharded_ddp

Notes:

> - This feature requires distributed training (so multiple GPUs)
> - It is not implemented for TPUs
> - It works with ``--fp16`` too, to make things even faster.
> - One of the main benefits of enabling ``--sharded_ddp`` is that it uses a lot less GPU memory, so you should be able

Reviewer 2: ```suggestion
- This feature requires distributed training (so multiple GPUs).
- It is not implemented for TPUs.
- It works with ``--fp16`` too, to make things even faster.
- One of the main benefits of enabling ``--sharded_ddp`` is that it uses a lot less GPU memory, so you should be able
```

   .. code-block:: bash

       deepspeed --num_gpus=2 your_program.py <normal cl args> --deepspeed ds_config.json

   Unlike, ``torch.distributed.launch`` where you have to specify how many GPUs to use with ``--nproc_per_node``, with
   the ``deepspeed`` launcher you don't have to use the corresponding ``--num_gpus`` if you want all of your GPUs used.
   The full details on how to configure various nodes and GPUs can be found `here
   <https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node>`__.

   Here is an example of running ``finetune_trainer.py`` under DeepSpeed deploying all available GPUs:

   .. code-block:: bash

       cd examples/seq2seq
       deepspeed ./finetune_trainer.py --deepspeed ds_config.json \
       --model_name_or_path sshleifer/distill-mbart-en-ro-12-4 --data_dir wmt_en_ro \
       --output_dir output_dir --overwrite_output_dir \
       --do_train --n_train 500 --num_train_epochs 1 \
       --per_device_train_batch_size 1  --freeze_embeds \
       --src_lang en_XX --tgt_lang ro_RO --task translation

   Note that in the DeepSpeed documentation you are likely to see ``--deepspeed --deepspeed_config ds_config.json`` -

>    i.e. 2 DeepSpeed-related arguments, but for simplicity-sake, and since there are already so many arguments to deal

Reviewer 2: ```suggestion
   i.e. two DeepSpeed-related arguments, but for the sake of simplicity, and since there are already so many arguments to deal
```

   with, we combined the two into a single argument.

> But before you can deploy DeepSpeed, let's discuss its configuration.

Reviewer 2: ```suggestion
Before you can deploy DeepSpeed, let's discuss its configuration.
```

**Configuration:**

For the complete guide to the DeepSpeed configuration options that can be used in its configuration file please refer
to the `following documentation <https://www.deepspeed.ai/docs/config-json/>`__.

While you always have to supply the DeepSpeed configuration file, you can configure the DeepSpeed integration in
several ways:

1. Supply most of the configuration inside the file, and just use a few required command line arguments. This is the
   recommended way as it puts most of the configuration params in one place.
2. Supply just the ZeRO configuration params inside the file, and configure the rest using the normal Trainer command
   line arguments.

> 3. Any variation of the first two ways

Reviewer 2: ```suggestion
3. Any variation of the first two ways.
```

To get an idea of what DeepSpeed configuration file looks like, here is one that activates ZeRO stage 2 features,

> enables fp16, uses Adam optimizer and WarmupLR scheduler:

Reviewer 2: ```suggestion
enables FP16, uses Adam optimizer and WarmupLR scheduler:
```

.. code-block:: json

    {
        "fp16": {
            "enabled": true,
            "loss_scale": 0,
            "loss_scale_window": 1000,
            "hysteresis": 2,
            "min_loss_scale": 1
        },

       "zero_optimization": {
           "stage": 2,
           "allgather_partitions": true,
           "allgather_bucket_size": 5e8,
           "overlap_comm": true,
           "reduce_scatter": true,
           "reduce_bucket_size": 5e8,
           "contiguous_gradients": true,
           "cpu_offload": true
       },

       "optimizer": {
         "type": "Adam",
         "params": {
           "lr": 3e-5,
           "betas": [
             0.8,
             0.999
           ],
           "eps": 1e-8,

>            "weight_decay": 3e-7

Reviewer 2: This example might be confusing as we use AdamW everywhere in the lib by default, so I would adjust it to use that (with a weight decay of 0.01 probably)

         }
       },

       "scheduler": {
         "type": "WarmupLR",
         "params": {
           "warmup_min_lr": 0,
           "warmup_max_lr": 3e-5,
           "warmup_num_steps": 500
         }
       }
    }

> If you already have a command line that you have been using with HF Trainer args, you can continue using those and the

Reviewer 2: ```suggestion
If you already have a command line that you have been using with :class:`transformers.Trainer` args, you can continue using those and the
```
(no need to say HF Trainer if the Trainer is linked).

Trainer will automatically convert them into the corresponding DeepSpeed configuration at run time. For example, you
could use the following configuration file:

.. code-block:: json

    {
       "zero_optimization": {
           "stage": 2,
           "allgather_partitions": true,
           "allgather_bucket_size": 5e8,
           "overlap_comm": true,
           "reduce_scatter": true,
           "reduce_bucket_size": 5e8,
           "contiguous_gradients": true,
           "cpu_offload": true
       }
    }

and the following command line arguments:

.. code-block:: bash

    --learning_rate 3e-5 --warmup_steps 500 --adam_beta1 0.8 --adam_beta2 0.999 --adam_epsilon 1e-8 \

>     --weight_decay 3e-7 --lr_scheduler_type constant_with_warmup --fp16 --fp16_backend amp

Reviewer 2: Trainer uses AdamW by default and not Adam. Does Trainer with DeepSpeed changes that? It should be documented if that's the case.

.. autoclass:: transformers.TFTrainingArguments
    :members:

Trainer Integrations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The :class:`~transformers.Trainer` has been extended to support libraries that may dramatically improve your training
time and fit much bigger models.

> Currently it support third party solutions, `DeepSpeed <https://github.com/microsoft/DeepSpeed>`__ and `FairScale

Reviewer 3: ```suggestion
Currently it supports third party solutions, `DeepSpeed <https://github.com/microsoft/DeepSpeed>`__ and `FairScale
```

   or find more details on `the FairScale's github page
   <https://github.com/facebookresearch/fairscale/#installation>`__.

2. Add ``--sharded_ddp`` to the command line arguments, and make sure you have added the distributed launcher ``-m
   torch.distributed.launch --nproc_per_node=NUMBER_OF_GPUS_YOU_HAVE`` if you haven't been using it already.

For example here is how you could use it for ``finetune_trainer.py`` with 2 GPUs:

.. code-block:: bash

    cd examples/seq2seq
    python -m torch.distributed.launch --nproc_per_node=2 ./finetune_trainer.py \
    --model_name_or_path sshleifer/distill-mbart-en-ro-12-4 --data_dir wmt_en_ro \
    --output_dir output_dir --overwrite_output_dir \
    --do_train --n_train 500 --num_train_epochs 1 \
    --per_device_train_batch_size 1  --freeze_embeds \
    --src_lang en_XX --tgt_lang ro_RO --task translation \
    --fp16 --sharded_ddp

Notes:

- This feature requires distributed training (so multiple GPUs).
- It is not implemented for TPUs.
- It works with ``--fp16`` too, to make things even faster.
- One of the main benefits of enabling ``--sharded_ddp`` is that it uses a lot less GPU memory, so you should be able

>   to use significantly larger batch sizes using the same hardware (e.g. 3x and even bigger) which should leave to

Reviewer 3: ```suggestion
  to use significantly larger batch sizes using the same hardware (e.g. 3x and even bigger) which should lead to
```

which ZeRO stages you want to enable and how to configure them.

.. code-block:: json

    {
       "zero_optimization": {
           "stage": 2,
           "allgather_partitions": true,
           "allgather_bucket_size": 5e8,
           "overlap_comm": true,
           "reduce_scatter": true,
           "reduce_bucket_size": 5e8,
           "contiguous_gradients": true,
           "cpu_offload": true
       }
    }

Notes:

- enabling ``cpu_offload`` should reduce GPU RAM usage (it requires ``"stage": 2``)
- ``"overlap_comm": true`` trades off increased GPU RAM usage to lower all-reduce latency. ``overlap_comm`` uses 4.5x
  the ``allgather_bucket_size`` and ``reduce_bucket_size`` values. So if they are set to 5e8, this requires a 9GB
  footprint (``5e8 x 2Bytes x 2 x 4.5``). Therefore, if you have a GPU with 8GB or less RAM, to avoid getting
  OOM-errors you will need to reduce those parameters to about ``2e8``, which would require 3.6GB.

This section has to be configured exclusively via DeepSpeed configuration - the :class:`~transformers.Trainer` provides
no equivalent command line arguments.

**Optimizer:**

> DeepSpeed has several tested with ZeRO optimizers, which are Adam, OneBitAdam, and Lamb. It, however, can import other

Reviewer 3: I don't understand the first sentence

Author: It means that "It has tested only these optimizers to properly work with ZeRO". I will rewrite it not to use passive and it will be then straightforward.

Author: I rewrote it as:

> DeepSpeed's main optimizers are Adam, OneBitAdam, and Lamb. These have been thoroughly tested with ZeRO and are thus
recommended to be used. It, however, can import other optimizers from ``torch``. 

Please let me know if it's still unclear.

> DeepSpeed has several tested with ZeRO optimizers, which are Adam, OneBitAdam, and Lamb. It, however, can import other

Reviewer 3: Maybe you meant
```suggestion
DeepSpeed has several tested ZeRO optimizers, which are Adam, OneBitAdam, and Lamb. It, however, can import other
```
?

Author: see above.
